### PR TITLE
Flags support

### PR DESCRIPTION
**Pull request by paralainer**
Added support of memcached flags

### CODE REVIEW

File: folsom/src/main/java/com/spotify/folsom/MemcacheClient.java [hunk 1]
```diff
@@ -104,14 +153,25 @@
...
   CompletionStage<GetResult<V>> casGet(String key);
 
+  /**
+   * Get the value for the provided key, including the CAS value and flags
+   *
+   * @param key First key, must not be null
+   * @return A future representing completion of the request, with the value, including the CAS
+   *     value and flags, or null if the value does not exists.
+   */
+  default CompletionStage<GetResult<V>> getWithFlags(String key) {
```
Comment (paralainer) [inline folsom/src/main/java/com/spotify/folsom/MemcacheClient.java:169]: This is basically a copy of `casGet`, it looks better when we have a method that explicitly says that it returns flags as well.

File: folsom/src/main/java/com/spotify/folsom/client/ascii/AsciiMemcacheDecoder.java [hunk 2]
```diff
@@ -179,6 +183,7 @@ protected void decode(final ChannelHandlerContext ctx, final ByteBuf buf, final
             this.value = new byte[size];
             this.valueOffset = 0;
             this.cas = cas;
+            this.flags = new Flags(flags);
```
Comment (spkrka) [inline folsom/src/main/java/com/spotify/folsom/client/ascii/AsciiMemcacheDecoder.java:186]: If we store this as a primitive int, we can avoid some unnecessary garbage creation

File: folsom/src/main/java/com/spotify/folsom/client/binary/BinaryMemcacheDecoder.java [hunk 3]
```diff
@@ -66,7 +67,12 @@ protected void decode(final ChannelHandlerContext ctx, final ByteBuf buf, final
         return;
       }
 
-      buf.skipBytes(extrasLength);
+      Flags flags = Flags.DEFAULT;
+      if (extrasLength > 0) {
+        flags = new Flags(buf.readInt());
```
Comment (spkrka) [inline folsom/src/main/java/com/spotify/folsom/client/binary/BinaryMemcacheDecoder.java:72]: use primitive instead of allocating?

**Comment by spkrka** [comment]
I think if we want to introduce flags, we should either do it high level or low level:

High-level: Introduce a proper enumset of valid flags that are supported
Low-level: use a primitive int that works with the API

Right now this PR is mostly low-level, but with an int-wrapper class. I think we can remove that wrapper class.

**Comment by paralainer** [comment]
> High-level

I don't think it makes sense as it is pretty much an arbitrary value and it's up to consumer to decide which flags they need.

> Right now this PR is mostly low-level, but with an int-wrapper class

To be honest the only reason I introduced Flags class is that `int` interfere with other parameters which are `long`, it broke some tests and will most likely break some consumers code if they rely on automatic `int -> long` conversion.
Other approach could be instead of reusing the same method names create new methods with `withFlags` suffix, like `setWithFlags`. 

On the point of allocations, if people won't use flags they won't have allocations as it will use the `Flags.DEFAULT` which is static (I can make an optimization on response side as well, so if flags are 0 we use `Flags.DEFAULT`).

Having `Flags` class we can implement some convenience methods in it later on, which can help with bitmasks for instance. Also if you use this library from Kotlin it make it easier to build extension methods.

**Pull request closed without merging**